### ISSUE TITLE
Question: How about Grok?

### ISSUE THREAD

**Issue by PetrKotynek** [ruven/iipsrv#265]
Hi,

I quote from iipsrv 1.1 released:
"A possible alternative to OpenJPEG is Grok, an open source fork of OpenJPEG that is a drop-in replacement for OpenJPEG and which includes additional optimizations."

I tried IIPImage Server 1.2 in combination with same Grok versions but failed even with
openjpeg-like versions 2.x. This path ( drop-in replacement ) is definitely not easy at the moment

./configure --with-openjpeg=

 because Grok split  from openjpeg a long time ago.  It's offering

./configure --with-grok=

It could be very fast and at the same time with the benefits of open source. 

Regards,

Petr Kotynek
Library of the Czech Academy of Sciences

**Comment by ruven** [comment]
Grok was originally a fork of OpenJPEG with a compatible API. However, Grok has since evolved and now has it's own completely separate API, so would need a new handler within iipsrv. There may well be a small speed improvement, but I'm not sure it will be so significant as OpenJPEG has also made good speed improvements in the last few releases.

If you really want to use Grok, you could try adapting the source code in [OpenJPEGImage.cc](https://github.com/ruven/iipsrv/blob/master/src/OpenJPEGImage.cc) to follow the Grok API. I'll put this on a list of future features to add, but can't make any guarantees about when this will be available (certainly not in the next release).

**Comment by bezverec** [comment]
We woud also welcome Grok implementation due to better performance compared to OpenJPEG, at least when measured outside image server environments. This is just me letting you know there is an ongoing interest in this, no pressure. Thank you for your great work! :)

**Comment by ruven** [comment]
OK, I'll keep this in mind. Have you done any benchmarks? How much faster is it compared to OpenJPEG? How does it compare to Kakadu?

**Comment by bezverec** [comment]
here is one of my custom benchmark results (the benchmark itself is by no means perfect).

[bench2000_report_2026-02-03_12-45-11.pdf](https://github.com/user-attachments/files/26244308/bench2000_report_2026-02-03_12-45-11.pdf)

The performance of Grok codec seems to be much closer to Kakadu than OpenJPEG. I am unable to guarantee similar results under different conditions, though. However in Czech cultural heritage digitization projects, Kakadu is still the golden standard (for those who can afford it). Some of us use OpenJPEG, but we are slowly shifting our attention towards Grok.

**Comment by ruven** [comment]
> here is one of my custom benchmark results (the benchmark itself is by no means perfect).

Interesting! That's much closer to Kakadu than I expected.

Another option that I've been looking at that could potentially beat Kakadu would be this Nvidia JPEG2000 codec: https://docs.nvidia.com/cuda/nvjpeg2000/index.html. This would, of course, require your server to have an Nvidia GPU, but it could in theory make JPEG2000 as fast as TIFF.

**Comment by bezverec** [comment]
Interesting! Thanks for sharing. There is also JasPer, but I have not been able to test that one yet.

**Comment by sbarex** [comment]
I tried using Grok for jp2000 files. I used some AI to create the code to use Grok instead of OpenJpeg, so there is probably room for improvement.

iipsrv was launched from terminal with:
```sh
CORS=* LOGFILE=/tmp/iipsrv.log MAX_IMAGE_CACHE_SIZE=0 JPEG_QUALITY=90 MAX_CVT=3000 iipsrv.fcgi --bind 127.0.0.1:9006
```

The webserver is nginx, the system is a MacBook Pro M4Pro with 48Gb Ram.

The test image is 14.182 x 24.815 px, I tried to run a benchmark with `ab` which requires the entire image. 

## Whole image extraction test
(Output automatically reduced to 3.000px due to MAX_CVT env)

Benchmarck with 100 calls (10 concurrencies):
```sh
ab -n 100 -c 10 http://iipsrv.test/fcgi-bin/iipsrv.fcgi?IIIF=image.jp2/full/full/0/default.jpg
```

|format|time|memory|
|---|---:|---:|
|tif|15s|~ 40mb|
|jp2 Grok (all available threads)|43s|~ 1Gb|
|jp2 Grok (only 2 threads)|72s|~ 200mb|
|jp2 OpenJpeg|116s|~ 110mb|

# Single tile extraction test
Benchmarck with 500 calls (10 concurrencies):

```sh
ab -n 500 -c 10 http://iipsrv.test/fcgi-bin/iipsrv.fcgi?FIF=image.jp2&JTL=7,100
```

format|time|memory|
|---|---:|---:|
|tif|0.6s|~ 4mb|
|jp2 Grok (all available threads)|3.5s|~ 22mb|
|jp2 Grok (only 2 threads)|8.3s|~ 22mb|
|jp2 OpenJpeg|29s|~ 30mb|

**Comment by bezverec** [comment]
Thanks for sharing. I have done some more benchmarking recently. Full disclosure: my results were mixed, Grok and OpenJPEG were similarly fast in my scenarios - both imageserver and local machine encoding/decoding, but closer to Kakadu than I expected. But it is often difficult for me do discern correct comparable parameters (the defaults and flags are quite different for each codec). In my benchmark above I probably used wrong parameters, especially for OpenJPEG putting it to disadvantage.

**Comment by sbarex** [comment]
> Thanks for sharing. I have done some more benchmarking recently. Full disclosure: my results were mixed, Grok and OpenJPEG were similarly fast in my scenarios - both imageserver and local machine encoding/decoding, but closer to Kakadu than I expected. But it is often difficult for me do discern correct comparable parameters (the defaults and flags are quite different for each codec). In my benchmark above I probably used wrong parameters, especially for OpenJPEG putting it to disadvantage.

I have update the benchmark with the tile extraction test.

**Comment by ruven** [comment]
> I used some AI to create the code to use Grok instead of OpenJpeg

Is this something we could use as the basis for a full implementation? If not, the documentation for Grok is rather thin, but there is the GDAL implementation we could use as a basis (https://github.com/OSGeo/gdal/blob/master/frmts/grok/grkdatasetbase.h) or the code from the Grok command-line decoder.

**Comment by sbarex** [comment]
Of course, I haven't extensively verified its functionality or the presence of memory leaks.
I'm not very good at forking and PR, so I'm attaching the files I modified here.

I compile/test the source on a macOS 26.5 arm M4 Pro, grok installed with homebrew:

```bash
./autogen.sh

./configure CFLAGS="-I/opt/homebrew/include -I/opt/homebrew/include/grok" LDFLAGS="-L/opt/homebrew/lib -L/opt/homebrew/Cellar/grokj2k/20.3.6/lib -lgrokj2k" CXXFLAGS="-I/opt/homebrew/include -I/opt/homebrew/include/grok" --prefix=/Users/sbarex/iipsrv_grok/buil

make

make install
```

then I run with:

```bash
CORS=* LOGFILE=/tmp/iipsrv.log MAX_IMAGE_CACHE_SIZE=0 JPEG_QUALITY=90 MAX_CVT=3000 /Users/sbarex/iipsrv_grok/build/sbin/iipsrv --bind 127.0.0.1:9006
```

[iipsrv_grok.zip](https://github.com/user-attachments/files/30281238/iipsrv_grok.zip)

**Comment by bezverec** [comment]
Interesting, thanks. I used your code and tried to move it further. The results is in PR #290. It is AI coded. But perhpas can be somewhat useful if not fit to be merged.
